staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go

```diff
@@ -279,26 +282,43 @@ func BenchmarkShuffleAndDealToSlice(b *testing.B) {
 }
 
 func TestUniformDistribution(t *testing.T) {
-	deckSize, handSize := int32(128), int32(3)
-	handCoordinateMap := make(map[int]int)
+	var aHand [3]int
+	deckSize := int32(128)
+	handCoordinateMap := make(map[int]int) // maps coded hand to count of times seen
 
-	allCoordinateCount := 128 * 127 * 126 / 6
+	fallingFactorial := 128 * 127 * 126
+	allCoordinateCount := fallingFactorial / 6
+	hashMax := uint64(16 << uint32(math.Ceil(math.Log2(float64(fallingFactorial)))))
+	center := float64(hashMax) / float64(allCoordinateCount)
 
-	for i := 0; i < allCoordinateCount*16; i++ {
-		hands := ShuffleAndDealToSlice(rand.Uint64(), deckSize, handSize)
-		sort.Slice(hands, func(i, j int) bool {
-			return hands[i] < hands[j]
-		})
+	for i := uint64(0); i < hashMax; i++ {
+		ShuffleAndDealIntoHand(i, deckSize, aHand[:])
+		sort.IntSlice(aHand[:]).Sort()
 		handCoordinate := 0
-		for _, hand := range hands {
-			handCoordinate = handCoordinate<<7 + int(hand)
+		for _, card := range aHand {
+			handCoordinate = handCoordinate<<7 + card
 		}
 		handCoordinateMap[handCoordinate]++
 	}
 
-	// TODO: check uniform distribution
-	t.Logf("%d", len(handCoordinateMap))
-	t.Logf("%d", allCoordinateCount)
+	t.Logf("Number of possible hands = %d, number of hands seen = %d, number of deals = %d, central count = %v", allCoordinateCount, len(handCoordinateMap), hashMax, center)
+
+	// histogram maps (count of times a hand is seen) to (number of hands having that count)
+	histogram := make(map[int]int)
+	var centralSum int
+	for _, count := range handCoordinateMap {
+		histogram[count] = histogram[count] + 1
+	}
+
+	for count := int(math.Ceil(center * 0.9)); count <= int(math.Floor(center*1.1)); count++ {
```

Comment:
There are still 3 fixed number here for test, 0.9, 1.1 and 0.99, and they are linked, hard to maintain I think. How about using standard deviation? Then we will only have one fixed number.

Reply:
cc @yue9944882

Reply:
am fine with the hardcodes in the test

Reply:
There are really only two numbers, not three; 0.9 and 1.1 are just 0.1 below and above unity.

I am not sure there is a real issue with maintainability here.  The ShuffleAndDeal function is intended to implement a certain mathematical function; alternative implementations of the same function will produce exactly the same histogram (because it is a consequence of the mathematical function of shuffle-and-deal).

Reply:
Could we bikeshed the evaluation here in follow-up PRs?  I would like to get the validation function put to bed.

From the math of the ShuffleAndDeal function, when N*fallingFactorial (for any whole number N) consecutive hash values are studied we know that each hand will be dealt the same number of times and that number is N*factorial(handSize).  So to test that this is what happened the evaluation should sum the histogram from floor(hashMax/fallingFactorial)*factorial(handSize) through ceiling(hashMax/fallingFactorial)*factorial(handSize) and insist on findig 100% of the mass there.

Reply:
I revised again, making the uniformity tester run multiple test cases and evaluate whether all the counts in the histogram fall in the expected range.